src/DSharp.Compiler/Generator/ExpressionGenerator.cs

```diff
@@ -73,40 +73,54 @@ internal static class ExpressionGenerator
                     return;
                 }
 
-                if (expression.LeftOperand is IndexerExpression indexExpression &&
-                    !indexExpression.Indexer.UseScriptIndexer)
+                if (expression.LeftOperand is IndexerExpression indexExpression)
                 {
-                    Debug.Assert(indexExpression.Type == ExpressionType.Indexer);
+                    IndexerSymbol indexerSymbol = indexExpression.Indexer;
 
-                    if (indexExpression.ObjectReference is BaseExpression)
+                    if (!indexerSymbol.UseScriptIndexer)
                     {
-                        writer.Write(((BaseExpression) indexExpression.ObjectReference)
-                                     .EvaluatedType.FullGeneratedName);
-                        writer.Write(".prototype.set_");
-                        writer.Write(indexExpression.Indexer.GeneratedName);
-                        writer.Write(".call(");
-                        writer.Write(generator.CurrentImplementation.ThisIdentifier);
-                        writer.Write(", ");
-                        GenerateExpressionList(generator, symbol, indexExpression.Indices);
-                        writer.Write(", ");
-                        GenerateExpression(generator, symbol, expression.RightOperand);
-                        writer.Write(")");
+                        Debug.Assert(indexExpression.Type == ExpressionType.Indexer);
+
+                        if (indexExpression.ObjectReference is BaseExpression)
```

Comment:
We might as well do the cast herre to get the reference as a typed `BaseExpression` instead of casting it below :D